### PR TITLE
Handle deletion of profiles without a resync.

### PR DESCRIPTION
**Pull request by fasaxc**

### CODE REVIEW

**Comment by fasaxc** [comment]
Haha, you've been waiting for this one, eh? @Lukasa

**Comment by fasaxc** [comment]
We sanity checked this on the FV rig against one or two tests and it passed, I think ti's good to merge.

File: calico/datamodel_v1.py [hunk 1]
```diff
@@ -57,8 +57,12 @@
     r'/(?P<hostname>[^/]+)/.+/endpoint/(?P<endpoint_id>[^/]+)')
 
 
+def dir_for_host(hostname):
+    return VERSION_DIR + "/host/%s" % hostname
```
Comment (robbrockbank) [inline calico/datamodel_v1.py]: better as HOST_DIR + "/%s" %  hostname
?

File: calico/datamodel_v1.py [hunk 2]
```diff
@@ -57,8 +57,12 @@
...
+
+
 def dir_for_per_host_config(hostname):
-    return HOST_DIR + "/%s/config/" % hostname
+    return HOST_DIR + "/%s/config" % hostname
```
Comment (robbrockbank) [inline calico/datamodel_v1.py]: and maybe dir_for_host(hostname) + "/config"

File: calico/datamodel_v1.py [hunk 3]
```diff
@@ -80,3 +84,15 @@ def key_for_profile_tags(profile_id):
 
 def key_for_config(config_name):
     return CONFIG_DIR + "/%s" % config_name
+
+
+def get_profile_id_for_profile_dir(key):
+    """
+    :param str key: etcd key.
+    :returns The profile ID if this is a profile dir or None if not.
```
Comment (robbrockbank) [inline calico/datamodel_v1.py]: or False if the key isn't valid ?

Comment (robbrockbank) [inline calico/datamodel_v1.py]: or should that case also return None?

File: calico/felix/fetcd.py [hunk 4]
```diff
@@ -267,12 +288,13 @@ def watch_etcd(self, update_splitter):
...
-                if response.action not in ("set", "create"):
-                    # FIXME: this check is over-broad.
+                if (response.action not in ("set", "create") and
+                        any((response.key.startswith(pfx) for pfx in
+                             PREFIXES_TO_RESYNC_ON_CHANGE))):
                     # It's purpose is to catch deletions of whole directories
                     # or other operations that we're not expecting.
-                    _log.warning("Unexpected action %s to %s; triggering "
-                                 "resync.", response.action, response.key)
+                    _log.warning("Unexpected event: %s; triggering resync.",
```
Comment (robbrockbank) [inline calico/felix/fetcd.py]: should this really be a warning just for non-deletes?

File: calico/datamodel_v1.py (hunk 1 above)
Comment (fasaxc) [inline calico/datamodel_v1.py]: Yep

File: calico/datamodel_v1.py (hunk 2 above)
Comment (fasaxc) [inline calico/datamodel_v1.py]: Yep.

File: calico/datamodel_v1.py (hunk 3 above)
Comment (fasaxc) [inline calico/datamodel_v1.py]: Should return None.  Will fix.

File: calico/felix/fetcd.py (hunk 4 above)
Comment (fasaxc) [inline calico/felix/fetcd.py]: No, this case actually catches delete for, say a whole workload or a whole host right now.  In those cases, we can't yet handle it gracefully so we have to resync.

**Comment by robbrockbank** [comment]
looks good

**Comment by fasaxc** [comment]
Fixes #382

**Pull request merged**